x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.store;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.hash.MessageDigests;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * RoleReference is a handle to the actual role definitions (role descriptors).
+ * It has different sub-types depending on how the role descriptors should be resolved.
+ */
+public interface RoleReference {
+
+    /**
+     * The key for caching purpose. Callers can use this value to determine whether it should skip
+     * resolving the role descriptors and subsequently building the role.
+     */
+    RoleKey cacheKey();
```

Comment:
Do we explicitly want to call this a cache key? or simply an id?
Mostly I'm wondering if there's places where it would be useful to be able to describe a RoleReference (as an id), but maybe that's handled by `toString()`?

Calling it a cacheKey just feels artificially restrictive to me.

Reply:
I changed it to `id`. Your suggestion makes sense to me. Also I think it is better to avoid using the word `cache` in `RoleReference` since in theory it does not know and should not care about caching.